ext/mbstring/libmbfl/filters/mbfilter_ucs4.c

```diff
@@ -125,153 +147,146 @@ const struct mbfl_convert_vtbl vtbl_wchar_ucs4le = {
-		n = (c & 0xff) << 16;
-		filter->cache |= n;
-	} else if (filter->status == 2) {
-		filter->status = 3;
-		n = (c & 0xff) << 8;
-		filter->cache |= n;
+	filter->cache = (filter->cache << 8) | (c & 0xFF);
+	if (filter->status < 3) {
+		filter->status++;
 	} else {
-		filter->status = 0;
-		n = (c & 0xff) | filter->cache;
-		CK((*filter->output_function)(n, filter->data));
+		CK(emit_wchar_if_valid(filter->cache, filter));
+		filter->status = filter->cache = 0;
 	}
 	return c;
 }
 
-/*
- * wchar => UCS-4BE
- */
 int mbfl_filt_conv_wchar_ucs4be(int c, mbfl_convert_filter *filter)
 {
 	if (c >= 0 && c < MBFL_WCSGROUP_UCS4MAX) {
-		CK((*filter->output_function)((c >> 24) & 0xff, filter->data));
-		CK((*filter->output_function)((c >> 16) & 0xff, filter->data));
-		CK((*filter->output_function)((c >> 8) & 0xff, filter->data));
-		CK((*filter->output_function)(c & 0xff, filter->data));
+		CK((*filter->output_function)((c >> 24) & 0xFF, filter->data));
+		CK((*filter->output_function)((c >> 16) & 0xFF, filter->data));
+		CK((*filter->output_function)((c >> 8) & 0xFF, filter->data));
+		CK((*filter->output_function)(c & 0xFF, filter->data));
 	} else {
 		CK(mbfl_filt_conv_illegal_output(c, filter));
 	}
 
 	return c;
 }
 
-/*
- * UCS-4LE => wchar
- */
 int mbfl_filt_conv_ucs4le_wchar(int c, mbfl_convert_filter *filter)
 {
-	int n;
-
-	if (filter->status == 0) {
-		filter->status = 1;
-		n = (c & 0xff);
-		filter->cache = n;
-	} else if (filter->status == 1) {
-		filter->status = 2;
-		n = (c & 0xff) << 8;
-		filter->cache |= n;
-	} else if (filter->status == 2) {
-		filter->status = 3;
-		n = (c & 0xff) << 16;
-		filter->cache |= n;
+	filter->cache = (filter->cache >> 8) | ((c & 0xFF) << 24);
+	if (filter->status < 3) {
+		filter->status++;
 	} else {
-		filter->status = 0;
-		n = ((c & 0xffu) << 24) | filter->cache;
-		CK((*filter->output_function)(n, filter->data));
+		CK(emit_wchar_if_valid(filter->cache, filter));
+		filter->status = filter->cache = 0;
 	}
 	return c;
 }
 
-/*
- * wchar => UCS-4LE
- */
 int mbfl_filt_conv_wchar_ucs4le(int c, mbfl_convert_filter *filter)
 {
 	if (c >= 0 && c < MBFL_WCSGROUP_UCS4MAX) {
-		CK((*filter->output_function)(c & 0xff, filter->data));
-		CK((*filter->output_function)((c >> 8) & 0xff, filter->data));
-		CK((*filter->output_function)((c >> 16) & 0xff, filter->data));
-		CK((*filter->output_function)((c >> 24) & 0xff, filter->data));
+		CK((*filter->output_function)(c & 0xFF, filter->data));
+		CK((*filter->output_function)((c >> 8) & 0xFF, filter->data));
+		CK((*filter->output_function)((c >> 16) & 0xFF, filter->data));
+		CK((*filter->output_function)((c >> 24) & 0xFF, filter->data));
 	} else {
 		CK(mbfl_filt_conv_illegal_output(c, filter));
 	}
 
 	return c;
 }
+
+static int mbfl_filt_ident_ucs4be(int c, mbfl_identify_filter *filter)
+{
+	int n = filter->status & 0xFF; /* # of bytes already consumed in code unit */
+	if (n < 3) {
+		filter->status = (filter->status + 1) | (c << (8 * (3 - n)));
```

Comment:
This is awkward. Maybe add a `cache` field to the identify filter as well? That would allow us to keep code between the conversion and the identify filters more symmetrical.

Reply:
Will do.